File: LQWaveCircleProgressView/LQWaveCircleProgressView.m
//
//  LQWaveCircleProgressView.m
//  TestDemo
//
//  Created by hongzhiqiang on 2018/10/22.
//  Copyright © 2018 hhh. All rights reserved.
//

#import "LQWaveCircleProgressView.h"

@interface LQWaveCircleProgressView ()

@property (nonatomic, strong) CAShapeLayer * waveSinLayer;
@property (nonatomic, assign) NSInteger phase;
@property (nonatomic, strong) UIView * waveSinView;
@property (nonatomic, strong) UILabel *percentL;

@end

@implementation LQWaveCircleProgressView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor grayColor];
    }
    return self;
}

- (void)startLoadView {
    [self setupView];
}

- (void)setupView {
    [self addSubview:self.waveSinView];
    self.waveSinView.layer.mask = self.waveSinLayer;
    [self addSubview:self.percentL];
    CGFloat minRadius = MIN(self.bounds.size.width, self.bounds.size.height);
    self.layer.cornerRadius = minRadius/2;
    self.layer.masksToBounds = YES;
    CADisplayLink *displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(updateUI:)];
    [displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
}

- (void)setProgress:(CGFloat)progress {
    _progress = progress;
    if (progress > 1.0) {
        return;
    }
    [self creatAnimation];
}

- (void)setTextColor:(UIColor *)textColor {
    _textColor = textColor;
    if (textColor) {
        self.percentL.textColor = textColor;
    }
}

- (void)setWaveColor:(UIColor *)waveColor {
    _waveColor = waveColor;
    if (waveColor) {
        self.waveSinView.backgroundColor = waveColor;
    }
}

- (void)setTextFont:(UIFont *)textFont {
    _textFont = textFont;
    if (textFont) {
        self.percentL.font = textFont;
    }
}


- (void)creatAnimation {
    
    CGPoint position = CGPointMake(self.waveSinLayer.position.x, (1.5-_progress)*self.bounds.size.height);
    CGPoint fromPosition = self.waveSinLayer.position;
    self.waveSinLayer.position = position;
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"position"];
    animation.fromValue = [NSValue valueWithCGPoint:fromPosition];
    animation.toValue = [NSValue valueWithCGPoint:position];
    animation.duration = 0.25;
    animation.removedOnCompletion = NO;
    animation.fillMode = kCAFillModeForwards;
    [self.waveSinLayer addAnimation:animation forKey:@"positionWave"];
    
    __weak typeof(self) wself = self;
    [UIView animateWithDuration:0.25 animations:^{
        wself.percentL.center = CGPointMake(wself.bounds.size.width/2, wself.waveSinLayer.position.y-wself.bounds.size.height/2);
    } completion:^(BOOL finished) {
        wself.percentL.text = [NSString stringWithFormat:@"%.f%@",wself.progress*100,@"%"];
    }];
}

- (void)updateUI:(CADisplayLink *)displayLink {
    CGFloat scale = self.bounds.size.width/200;
    self.phase += 5*scale;
    self.waveSinLayer.path = [self path].CGPath;
}

- (UIBezierPath *)path {
    UIBezierPath *path = [UIBezierPath bezierPath];
    CGFloat width =  CGRectGetWidth(self.bounds);
    CGFloat height = CGRectGetHeight(self.bounds);
    CGFloat lastX = 0;
    for (CGFloat x = 0; x < width+1; x++) {
        CGFloat y = height*0.05*sin(2*M_PI/width*x*0.8+self.phase*2*M_PI/width)+height*0.05;
        if (x == 0) {
            [path moveToPoint:CGPointMake(x, y)];
        }
        else {
            [path addLineToPoint:CGPointMake(x, y)];
        }
        lastX = x;
        if (self.progress >= 1.0) {
           self.percentL.center = CGPointMake( self.percentL.center.x, self.bounds.size.height/2+y/2);
        }
        else {
            self.percentL.center = CGPointMake( self.percentL.center.x, self.waveSinLayer.position.y-self.bounds.size.height/2+y/2);
        }
        
    }
    
    [path addLineToPoint:CGPointMake(lastX, height)];
    [path addLineToPoint:CGPointMake(0, height)];
    return path;
}

#pragma mark - lazy

- (CAShapeLayer *)waveSinLayer {
    if (!_waveSinLayer) {
        CGFloat scale = self.bounds.size.width/200;
        _waveSinLayer = [CAShapeLayer new];
        _waveSinLayer.frame = CGRectMake(0, CGRectGetHeight(self.bounds)-10*scale, CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
        _waveSinLayer.backgroundColor = [UIColor clearColor].CGColor;
    }
    return _waveSinLayer;
}

- (UIView *)waveSinView {
    if (!_waveSinView) {
        _waveSinView = [UIView new];
        _waveSinView.frame = CGRectMake(0, 0, CGRectGetWidth(self.bounds), CGRectGetHeight(self.bounds));
        _waveSinView.backgroundColor = [UIColor colorWithRed:41/255.0 green:240/255.0 blue:253/255.0 alpha:0.8];
    }
    return _waveSinView;
}

- (UILabel *)percentL {
    if (!_percentL) {
        _percentL = [UILabel new];
        _percentL.textColor = [UIColor whiteColor];
        _percentL.backgroundColor = [UIColor clearColor];
        _percentL.frame = CGRectMake(0, 0, self.bounds.size.width, 30);
        _percentL.textAlignment = NSTextAlignmentCenter;
        _percentL.text = [NSString stringWithFormat:@"%.f%@",self.progress*100,@"%"];
        _percentL.center = CGPointMake(self.bounds.size.width/2, self.bounds.size.height - 10);
    }
    return _percentL;
}

@end

File: LQWaveCircleProgressView/LQWaveCircleProgressView.swift
//
//  LQWaveCircleProgressView.swift
//  TestDemo_Swift
//
//  Created by hongzhiqiang on 2018/10/23.
//  Copyright © 2018 hhh. All rights reserved.
//

import UIKit

class LQWaveCircleProgressView: UIView {
    
    public var progress : CGFloat = 0{
        didSet {
            if progress > 1.0 {
                return
            }
            self.creatAnimation()
        }
    }
    
    //水浪颜色,默认为浅蓝色
    public var waveColor : UIColor? {
        didSet {
            self.waveSinView.backgroundColor = waveColor
        }
    }
    
    //字体颜色,默认为白色
    public var textColor : UIColor? {
        didSet {
            self.percentL.textColor = textColor
        }
    }
    
    //字体大小,默认为系统默认大小
    public var textFont : UIFont? {
        didSet {
            self.percentL.font = textFont
        }
    }
    
    public func startLoadView() {
        self.setupView()
    }
    
    private var phase : CGFloat = 0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = .gray
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
   private func setupView() {
        self.addSubview(self.waveSinView)
        self.waveSinView.layer.mask = self.waveSinLayer
        let minRadius = min(self.bounds.size.width, self.bounds.size.height)
        self.layer.cornerRadius = minRadius/2
        self.layer.masksToBounds = true
        self.addSubview(self.percentL)
        
        let displayLink = CADisplayLink.init(target: self, selector: #selector(updateUI))
        displayLink.add(to: RunLoop.current, forMode: RunLoop.Mode.common)
    }
    
    @objc private func updateUI(){
        let scale = self.bounds.size.width / 200.0
        self.phase += 5.0*scale
        self.waveSinLayer.path = self.path().cgPath
    }
    
   private func path() -> UIBezierPath{
        let path = UIBezierPath()
        let width = self.bounds.size.width
        let height = self.bounds.size.width
        var lastX = 0
        
        for x in 0...Int(width) {
            let a = 2*Double.pi/Double(width)*Double(x)*0.8+Double(self.phase)*2*Double.pi/Double(width)
            let y = Double(height)*0.05*sin(a) + Double(height)*0.05
            if x == 0 {
                path.move(to: CGPoint(x: CGFloat(x), y: CGFloat(y)))
            }
            else{
                path.addLine(to: CGPoint(x: CGFloat(x), y: CGFloat(y)))
            }
            lastX = x
            
            if self.progress < 1.0 {
                self.percentL.center = CGPoint(x: self.bounds.size.width/2, y: self.waveSinLayer.position.y-self.bounds.size.height/2+CGFloat(y)/2)
            }
            else {
                self.percentL.center = CGPoint(x: self.bounds.size.width/2, y: self.bounds.size.height/2+CGFloat(y)/2)
            }
        }
        
        path.addLine(to: CGPoint(x: CGFloat(lastX), y: height))
        path.addLine(to: CGPoint(x: 0, y: height))
        return path
    }
    
   private func creatAnimation()  {
        let position = CGPoint(x: self.waveSinLayer.position.x, y: (1.5-self.progress)*self.bounds.size.height)
        let fromPosition = self.waveSinLayer.position
        self.waveSinLayer.position = position
        let animation = CABasicAnimation.init(keyPath: "position")
        animation.fromValue = NSValue(cgPoint: fromPosition)
        animation.toValue = NSValue(cgPoint: position)
        animation.duration = 0.25
        animation.isRemovedOnCompletion = false
        animation.fillMode = .forwards
        self.waveSinLayer.add(animation, forKey:nil)
        
        UIView.animate(withDuration: 0.25, animations: {
            self.percentL.center = CGPoint(x: self.bounds.size.width/2, y: self.waveSinLayer.position.y-self.bounds.size.height/2)
        }) { (finished) in
            self.percentL.text = String(format: "%.f%@", arguments: [self.progress*100,"%"])
        }
    }
    
   private lazy var waveSinLayer: CAShapeLayer = {
        let waveSinLayer = CAShapeLayer()
        let scale = self.bounds.size.width / 200
        waveSinLayer.frame = CGRect(x: 0, y:self.bounds.size.height-10*scale, width: self.bounds.size.width, height: self.bounds.size.height)
        waveSinLayer.backgroundColor = UIColor.clear.cgColor
        return waveSinLayer
    }()
    
   private lazy var waveSinView: UIView = {
        let waveSinView = UIView()
        waveSinView.frame = CGRect(x: 0, y: 0, width: self.bounds.size.width, height: self.bounds.size.height)
        waveSinView.backgroundColor = UIColor(red: 41/255.0, green: 240/255.0, blue: 253/255.0, alpha: 0.8)
        return waveSinView
    }()
    
   private lazy var percentL: UILabel = {
        let percentL = UILabel()
        percentL.textColor = .white
        percentL.backgroundColor = .clear
        percentL.frame = CGRect(x: 0, y: 0, width: self.bounds.size.width, height: 30)
        percentL.textAlignment = .center
        percentL.center = CGPoint(x: self.bounds.size.width/2, y: self.bounds.size.height - 10)
        percentL.text = String(format: "%.f%@", arguments: [self.progress*100,"%"])
        return percentL
    }()
    
}
